File: packages/core/src/backtest.test.ts
import { describe, expect, test } from "bun:test";
import { type BacktestLeg, type BacktestSettlement, backtestPair } from "./backtest";

const HOUR = 3_600_000;
const DAY = 86_400_000;
// 2026-09-01T00:00:00Z, so bucket dates are readable.
const START = Date.UTC(2026, 8, 1);

const series = (
  startMs: number,
  everyHours: number,
  count: number,
  rate: number,
): BacktestSettlement[] =>
  Array.from({ length: count }, (_, i) => ({
    settledAt: startMs + i * everyHours * HOUR,
    rate,
    basisHours: everyHours,
  }));

const leg = (
  venueId: string,
  venueSymbol: string,
  settlements: BacktestSettlement[],
): BacktestLeg => ({ venueId, venueSymbol, settlements });

describe("backtestPair", () => {
  test("sums each leg at its own cadence, and gets the signs the right way round", () => {
    // One day: hourly long leg paying 0.001% an hour, 8-hourly short leg receiving 0.01%.
    const result = backtestPair({
      long: leg("hyperliquid", "BTC", series(START, 1, 24, 0.00001)),
      short: leg("bybit", "BTCUSDT", series(START, 8, 3, 0.0001)),
      sizeUsd: 10_000,
      fromMs: START,
      toMs: START + DAY,
    });

    // Long pays 10000 * 0.00001 * 24 = 2.40; short receives 10000 * 0.0001 * 3 = 3.00.
    expect(result.long.fundingUsd).toBeCloseTo(-2.4, 9);
    expect(result.short.fundingUsd).toBeCloseTo(3, 9);
    expect(result.netFundingUsd).toBeCloseTo(0.6, 9);
    expect(result.long.settlements).toBe(24);
    expect(result.short.settlements).toBe(3);
    // 0.60 a day on 10k of notional is 0.006% a day, so 2.19% a year.
    expect(result.netFundingAprPercent).toBeCloseTo(2.19, 6);
  });

  test("a negative rate pays the long and charges the short", () => {
    const result = backtestPair({
      long: leg("gate", "X_USDT", series(START, 8, 3, -0.0002)),
      short: leg("okx", "X-USDT-SWAP", series(START, 8, 3, -0.0001)),
      sizeUsd: 1_000,
      fromMs: START,
      toMs: START + DAY,
    });

    expect(result.long.fundingUsd).toBeCloseTo(0.6, 9); // paid to hold the long
    expect(result.short.fundingUsd).toBeCloseTo(-0.3, 9); // charged to hold the short
    expect(result.netFundingUsd).toBeCloseTo(0.3, 9);
  });

  test("buckets by UTC day and counts winning days only among days that settled", () => {
    const long = [
      ...series(START, 8, 3, 0.0001), // day 1: long pays 3.00
      ...series(START + 2 * DAY, 8, 3, -0.0001), // day 3: long receives 3.00
    ];
    const result = backtestPair({
      long: leg("gate", "X_USDT", long),
      short: leg("okx", "X-USDT-SWAP", []),
      sizeUsd: 10_000,
      fromMs: START,
      toMs: START + 3 * DAY,
    });

    expect(result.perDay.map((day) => day.date)).toEqual(["2026-09-01", "2026-09-03"]);
    expect(result.perDay[0]?.netUsd).toBeCloseTo(-3, 9);
    expect(result.perDay[1]?.netUsd).toBeCloseTo(3, 9);
    // Day 2 had no settlement at all, so it is not counted either way.
    expect(result.winRateDays).toBeCloseTo(0.5, 9);
    expect(result.avgDailyUsd).toBeCloseTo(0, 9);
  });

  test("flags missing settlements instead of treating the gap as zero funding", () => {
    const withHole = [
      ...series(START, 8, 2, 0.0001),
      // 24h later: two settlements skipped.
      ...series(START + 32 * HOUR, 8, 2, 0.0001),
    ];
    const result = backtestPair({
      long: leg("gate", "X_USDT", withHole),
      short: leg("okx", "X-USDT-SWAP", series(START, 8, 4, 0.0001)),
      sizeUsd: 10_000,
      fromMs: START,
      toMs: START + 2 * DAY,
    });

    expect(result.long.missedSettlements).toBe(2);
    expect(result.long.missedHours).toBeCloseTo(16, 9);
    expect(result.short.missedSettlements).toBe(0);
  });

  test("survives an interval change mid-window without inventing settlements", () => {
    const changed = [
      ...series(START, 8, 3, 0.0001), // 8-hourly for a day
      ...series(START + DAY, 1, 24, 0.0000125), // then hourly
    ];
    const result = backtestPair({
      long: leg("bybit", "X", changed),
      short: leg("okx", "X", []),
      sizeUsd: 10_000,
      fromMs: START,
      toMs: START + 2 * DAY,
    });

    expect(result.long.settlements).toBe(27);
    expect(result.long.missedSettlements).toBe(0);
    expect(result.long.fundingUsd).toBeCloseTo(-(3 * 1 + 24 * 0.125), 9);
  });

  test("charges four taker fills and reports payback only when it repays", () => {
    const profitable = backtestPair({
      long: leg("a", "X", series(START, 8, 3, -0.0001)),
      short: leg("b", "X", series(START, 8, 3, 0.0001)),
      sizeUsd: 10_000,
      fromMs: START,
      toMs: START + DAY,
      fees: { longTakerBps: 5, shortTakerBps: 5 },
    });

    // 10000 * (5 + 5) / 10000 * 2 fills-per-leg = 20.
    expect(profitable.costsUsd).toBeCloseTo(20, 9);
    expect(profitable.netFundingUsd).toBeCloseTo(6, 9);
    expect(profitable.netAfterCostsUsd).toBeCloseTo(-14, 9);
    // 6/day against 20 of costs repays in 3 1/3 days.
    expect(profitable.paybackDays).toBeCloseTo(20 / 6, 9);

    const losing = backtestPair({
      long: leg("a", "X", series(START, 8, 3, 0.0001)),
      short: leg("b", "X", series(START, 8, 3, -0.0001)),
      sizeUsd: 10_000,
      fromMs: START,
      toMs: START + DAY,
      fees: { longTakerBps: 5, shortTakerBps: 5 },
    });
    expect(losing.paybackDays).toBeNull();
  });

  test("leaves costs null when fees are unknown, rather than assuming a number", () => {
    const result = backtestPair({
      long: leg("a", "X", series(START, 8, 3, -0.0001)),
      short: leg("b", "X", series(START, 8, 3, 0.0001)),
      sizeUsd: 10_000,
      fromMs: START,
      toMs: START + DAY,
    });

    expect(result.costsUsd).toBeNull();
    expect(result.netAfterCostsUsd).toBeNull();
    expect(result.paybackDays).toBeNull();
  });

  test("an empty window reports nothing rather than dividing by zero", () => {
    const result = backtestPair({
      long: leg("a", "X", []),
      short: leg("b", "X", []),
      sizeUsd: 10_000,
      fromMs: START,
      toMs: START,
    });

    expect(result.netFundingUsd).toBe(0);
    expect(result.netFundingAprPercent).toBe(0);
    expect(result.winRateDays).toBe(0);
    expect(result.perDay).toEqual([]);
  });
});

File: apps/collector-go/internal/core/backtest_test.go
package core

import (
	"math"
	"testing"
	"time"
)

// Vectors from packages/core/src/backtest.test.ts and the backtestPair cases of
// backtest-daily.test.ts. The side-by-side backtestDaily cases are not ported: there is no Go
// backtestDaily, because nothing in the collector calls it.

const (
	btHour = int64(3_600_000)
	btDay  = int64(86_400_000)
)

// 2026-09-01T00:00:00Z, so bucket dates are readable.
var btStart = time.Date(2026, 9, 1, 0, 0, 0, 0, time.UTC).UnixMilli()

// assertClose mirrors bun's toBeCloseTo(want, digits): |got - want| < 10^-digits / 2.
func assertClose(t *testing.T, label string, got, want float64, digits int) {
	t.Helper()
	if !(math.Abs(got-want) < math.Pow(10, -float64(digits))/2) {
		t.Errorf("%s: got %.15g, want %.15g (to %d digits)", label, got, want, digits)
	}
}

func btSeries(startMs int64, everyHours int64, count int, rate float64) []BacktestSettlement {
	out := make([]BacktestSettlement, count)
	for i := range out {
		out[i] = BacktestSettlement{
			SettledAt:  startMs + int64(i)*everyHours*btHour,
			Rate:       rate,
			BasisHours: float64(everyHours),
		}
	}
	return out
}

func btLeg(venueID, symbol string, settlements []BacktestSettlement) BacktestLeg {
	return BacktestLeg{VenueID: venueID, VenueSymbol: symbol, Settlements: settlements}
}

func TestBacktestSumsEachLegAtItsOwnCadence(t *testing.T) {
	result := BacktestPair(BacktestInput{
		Long:    btLeg("hyperliquid", "BTC", btSeries(btStart, 1, 24, 0.00001)),
		Short:   btLeg("bybit", "BTCUSDT", btSeries(btStart, 8, 3, 0.0001)),
		SizeUSD: 10_000,
		FromMs:  btStart,
		ToMs:    btStart + btDay,
	})
	// Long pays 10000 * 0.00001 * 24 = 2.40; short receives 10000 * 0.0001 * 3 = 3.00.
	assertClose(t, "long funding", result.Long.FundingUSD, -2.4, 9)
	assertClose(t, "short funding", result.Short.FundingUSD, 3, 9)
	assertClose(t, "net funding", result.NetFundingUSD, 0.6, 9)
	if result.Long.Settlements != 24 || result.Short.Settlements != 3 {
		t.Errorf("settlements %d/%d, want 24/3", result.Long.Settlements, result.Short.Settlements)
	}
	assertClose(t, "net APR", result.NetFundingAPRPercent, 2.19, 6)
}

func TestBacktestNegativeRatePaysTheLong(t *testing.T) {
	result := BacktestPair(BacktestInput{
		Long:    btLeg("gate", "X_USDT", btSeries(btStart, 8, 3, -0.0002)),
		Short:   btLeg("okx", "X-USDT-SWAP", btSeries(btStart, 8, 3, -0.0001)),
		SizeUSD: 1_000,
		FromMs:  btStart,
		ToMs:    btStart + btDay,
	})
	assertClose(t, "long funding", result.Long.FundingUSD, 0.6, 9)
	assertClose(t, "short funding", result.Short.FundingUSD, -0.3, 9)
	assertClose(t, "net funding", result.NetFundingUSD, 0.3, 9)
}

func TestBacktestBucketsByUTCDayAndCountsOnlySettledDays(t *testing.T) {
	long := append(btSeries(btStart, 8, 3, 0.0001), btSeries(btStart+2*btDay, 8, 3, -0.0001)...)
	result := BacktestPair(BacktestInput{
		Long:    btLeg("gate", "X_USDT", long),
		Short:   btLeg("okx", "X-USDT-SWAP", nil),
		SizeUSD: 10_000,
		FromMs:  btStart,
		ToMs:    btStart + 3*btDay,
	})
	if len(result.PerDay) != 2 || result.PerDay[0].Date != "2026-09-01" || result.PerDay[1].Date != "2026-09-03" {
		t.Fatalf("per day %+v, want 2026-09-01 and 2026-09-03", result.PerDay)
	}
	assertClose(t, "day 1", result.PerDay[0].NetUSD, -3, 9)
	assertClose(t, "day 3", result.PerDay[1].NetUSD, 3, 9)
	assertClose(t, "win rate", result.WinRateDays, 0.5, 9)
	assertClose(t, "avg daily", result.AvgDailyUSD, 0, 9)
}

func TestBacktestFlagsMissingSettlements(t *testing.T) {
	withHole := append(btSeries(btStart, 8, 2, 0.0001), btSeries(btStart+32*btHour, 8, 2, 0.0001)...)
	result := BacktestPair(BacktestInput{
		Long:    btLeg("gate", "X_USDT", withHole),
		Short:   btLeg("okx", "X-USDT-SWAP", btSeries(btStart, 8, 4, 0.0001)),
		SizeUSD: 10_000,
		FromMs:  btStart,
		ToMs:    btStart + 2*btDay,
	})
	if result.Long.MissedSettlements != 2 {
		t.Errorf("long missed %d, want 2", result.Long.MissedSettlements)
	}
	assertClose(t, "long missed hours", result.Long.MissedHours, 16, 9)
	if result.Short.MissedSettlements != 0 {
		t.Errorf("short missed %d, want 0", result.Short.MissedSettlements)
	}
}

func TestBacktestSurvivesAnIntervalChange(t *testing.T) {
	changed := append(btSeries(btStart, 8, 3, 0.0001), btSeries(btStart+btDay, 1, 24, 0.0000125)...)
	result := BacktestPair(BacktestInput{
		Long:    btLeg("bybit", "X", changed),
		Short:   btLeg("okx", "X", nil),
		SizeUSD: 10_000,
		FromMs:  btStart,
		ToMs:    btStart + 2*btDay,
	})
	if result.Long.Settlements != 27 || result.Long.MissedSettlements != 0 {
		t.Errorf("settlements %d missed %d, want 27 and 0", result.Long.Settlements, result.Long.MissedSettlements)
	}
	assertClose(t, "long funding", result.Long.FundingUSD, -(3*1 + 24*0.125), 9)
}

func TestBacktestChargesFourFillsAndPaybackOnlyWhenItRepays(t *testing.T) {
	fees := &BacktestFees{LongTakerBps: 5, ShortTakerBps: 5}
	profitable := BacktestPair(BacktestInput{
		Long:    btLeg("a", "X", btSeries(btStart, 8, 3, -0.0001)),
		Short:   btLeg("b", "X", btSeries(btStart, 8, 3, 0.0001)),
		SizeUSD: 10_000,
		FromMs:  btStart,
		ToMs:    btStart + btDay,
		Fees:    fees,
	})
	if profitable.CostsUSD == nil || profitable.NetAfterCostsUSD == nil || profitable.PaybackDays == nil {
		t.Fatal("costs, net after costs and payback must all be set when fees are known")
	}
	assertClose(t, "costs", *profitable.CostsUSD, 20, 9)
	assertClose(t, "net funding", profitable.NetFundingUSD, 6, 9)
	assertClose(t, "net after costs", *profitable.NetAfterCostsUSD, -14, 9)
	assertClose(t, "payback", *profitable.PaybackDays, 20.0/6, 9)

	losing := BacktestPair(BacktestInput{
		Long:    btLeg("a", "X", btSeries(btStart, 8, 3, 0.0001)),
		Short:   btLeg("b", "X", btSeries(btStart, 8, 3, -0.0001)),
		SizeUSD: 10_000,
		FromMs:  btStart,
		ToMs:    btStart + btDay,
		Fees:    fees,
	})
	if losing.PaybackDays != nil {
		t.Errorf("payback %v, want nil for a pair that never repays", *losing.PaybackDays)
	}
}

func TestBacktestLeavesCostsNilWhenFeesAreUnknown(t *testing.T) {
	result := BacktestPair(BacktestInput{
		Long:    btLeg("a", "X", btSeries(btStart, 8, 3, -0.0001)),
		Short:   btLeg("b", "X", btSeries(btStart, 8, 3, 0.0001)),
		SizeUSD: 10_000,
		FromMs:  btStart,
		ToMs:    btStart + btDay,
	})
	if result.CostsUSD != nil || result.NetAfterCostsUSD != nil || result.PaybackDays != nil {
		t.Error("costs must stay nil when fees are unknown")
	}
}

func TestBacktestEmptyWindowReportsNothing(t *testing.T) {
	result := BacktestPair(BacktestInput{
		Long:    btLeg("a", "X", nil),
		Short:   btLeg("b", "X", nil),
		SizeUSD: 10_000,
		FromMs:  btStart,
		ToMs:    btStart,
	})
	if result.NetFundingUSD != 0 || result.NetFundingAPRPercent != 0 || result.WinRateDays != 0 {
		t.Errorf("got net %v APR %v win %v, want all 0", result.NetFundingUSD, result.NetFundingAPRPercent, result.WinRateDays)
	}
	if result.PerDay == nil || len(result.PerDay) != 0 {
		t.Errorf("per day %#v, want an empty non-nil slice", result.PerDay)
	}
}

// dailyNets is a short leg netting `usd` on each successive day, one 8-hour settlement a day.
func dailyNets(nets []float64) BacktestResult {
	settlements := make([]BacktestSettlement, len(nets))
	for i, usd := range nets {
		settlements[i] = BacktestSettlement{SettledAt: btStart + int64(i)*btDay, Rate: usd / 10_000, BasisHours: 8}
	}
	return BacktestPair(BacktestInput{
		Long:    btLeg("okx", "BTC-USDT-SWAP", nil),
		Short:   btLeg("bybit", "BTCUSDT", settlements),
		SizeUSD: 10_000,
		FromMs:  btStart,
		ToMs:    btStart + int64(len(nets))*btDay,
	})
}

func TestBacktestBestWorstDayAndLargestFallFromAHigh(t *testing.T) {
	result := dailyNets([]float64{5, -8, 2, -1, 6})
	if result.BestDay == nil || result.BestDay.Date != "2026-09-05" {
		t.Fatalf("best day %+v, want 2026-09-05", result.BestDay)
	}
	assertClose(t, "best", result.BestDay.NetUSD, 6, 9)
	if result.WorstDay == nil || result.WorstDay.Date != "2026-09-02" {
		t.Fatalf("worst day %+v, want 2026-09-02", result.WorstDay)
	}
	assertClose(t, "worst", result.WorstDay.NetUSD, -8, 9)
	assertClose(t, "drawdown", result.MaxDrawdownUSD, 8, 9)
}

func TestBacktestOpeningLossDrawsDownFromZero(t *testing.T) {
	assertClose(t, "opening loss", dailyNets([]float64{-3, 1}).MaxDrawdownUSD, 3, 9)
	if got := dailyNets([]float64{1, 2, 3}).MaxDrawdownUSD; got != 0 {
		t.Errorf("rising drawdown %v, want 0", got)
	}
	empty := dailyNets(nil)
	if empty.BestDay != nil || empty.MaxDrawdownUSD != 0 {
		t.Errorf("empty best %+v drawdown %v, want nil and 0", empty.BestDay, empty.MaxDrawdownUSD)
	}
}

func TestBacktestAverageRateIsTimeWeightedAndSideless(t *testing.T) {
	settlements := make([]BacktestSettlement, 0, 17)
	for i := int64(0); i < 16; i++ {
		settlements = append(settlements, BacktestSettlement{SettledAt: btStart + i*btHour, Rate: 0.00001, BasisHours: 1})
	}
	settlements = append(settlements, BacktestSettlement{SettledAt: btStart + 16*btHour, Rate: 0.00008, BasisHours: 8})
	result := BacktestPair(BacktestInput{
		Long:    btLeg("okx", "BTC-USDT-SWAP", settlements),
		Short:   btLeg("bybit", "BTCUSDT", nil),
		SizeUSD: 10_000,
		FromMs:  btStart,
		ToMs:    btStart + btDay,
	})
	if !(result.Long.FundingUSD < 0) {
		t.Errorf("long funding %v, want negative", result.Long.FundingUSD)
	}
	if result.Long.AverageAPRPercent == nil {
		t.Fatal("long average APR is nil")
	}
	assertClose(t, "average APR", *result.Long.AverageAPRPercent, 8.76, 9)
	if result.Short.AverageAPRPercent != nil {
		t.Errorf("short average APR %v, want nil", *result.Short.AverageAPRPercent)
	}
}

func TestJSRoundTiesTowardPositiveInfinity(t *testing.T) {
	for _, c := range []struct{ in, want float64 }{
		{2.5, 3}, {-2.5, -2}, {1.4999, 1}, {-0.3, 0}, {-0.5, 0}, {-0.51, -1},
	} {
		if got := jsRound(c.in); got != c.want {
			t.Errorf("jsRound(%v) = %v, want %v", c.in, got, c.want)
		}
	}
}
